### ISSUE TITLE
Question about connecting taint flows

### ISSUE THREAD

**Issue by scottconstable** [github/codeql#13765]
I'm new to CodeQL, and I am trying to write a taint-tracking path query for C/C++ code. I have read the documentation here:
https://codeql.github.com/docs/writing-codeql-queries/creating-path-queries/#creating-path-queries
https://codeql.github.com/docs/codeql-language-guides/analyzing-data-flow-in-cpp-new/#analyzing-data-flow-in-cpp-new

I would like to write a query that finds a taint path with multiple steps, for example:
1. The taint source is a parameter to a subset of functions in the program
2. The tainted value must be used as an index to access a particular array before reaching the sink
3. The taint sink is a particular argument to a particular function

From the documentation above, I'm not sure how to write such a query. The `TaintTracking::Configuration::hasFlowPath()` API only takes a source and a sink. I don't know how to define the intermediate step. I looked at the `isAdditionalTaintStep()` API, but that doesn't seem to be what I need--from what I can tell, this API defines a new taint propagation rule, whereas I need a method to define a new constraint on the taint path.

Any help would be much appreciated.

**Comment by jketema** [comment]
Hi @scottconstable,

`isAdditionalTaintStep` should still be useable for this purpose, but want to use the version that has a so-called `FlowState`. You then probably want to have two possible states, one in which you saw the array access and one in which you didn't. In the `FlowState` version of `isSource` you then want to specify you're in the state in which you didn't see the access yet. In `isSink` you want to specify that you did see the access. And, `isAdditionalTaintStep` the two nodes should probably be equal, you want to check that the access occurs for the node, and you want to switch to one flow state to the other.

**Comment by scottconstable** [comment]
Hi @jketema,

Looking at the CodeQL documentation, I see that `FlowState` is provided by the `StateConfigSig` module: https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/ir/dataflow/internal/DataFlow.qll/module.DataFlow$StateConfigSig.html.

There is one method that does something like what you are describing: https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/ir/dataflow/internal/DataFlow.qll/predicate.DataFlow$isAdditionalFlowStep.2.html. But it looks like there is only such a method for dataflow analysis. I can't find a similar one for taint analysis.

**Comment by jketema** [comment]
The interface I was describing is here: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/tainttracking1/TaintTrackingImpl.qll#L160-L164


What you stumbled upon with `StateConfigSig` is the interface we're moving to, but which is not yet described in the documentation (this is currently being worked on). That interface can be used with both dataflow and taint flow.

If you want to use the new interface, which is already used in all the C++ queries in this repository, you can do:
```ql
module MyConfig implements DataFlow::StateConfigSig {
  class FlowState = ... ; // the type you want for the flow state, "boolean" would make sense in your case

  predicate isSource(DataFlow::Node node, FlowState flowstate) {
     // your usual isSource implementation
     and
     flowstate = false
  }

  predicate isSink(DataFlow::Node node, FlowState flowstate) {
     // your usual isSink implementation
     and
     flowstate = true
  }

  // Optionally implement any of the predicate from the signature that are marked as "default". For example:
  predicate isAdditionalFlowStep(
    DataFlow::Node node1, FlowState state1, DataFlow::Node node2, FlowState state2
  ) {
    // The additional step you had in mind that should change the state
   and
   state1 = false
   and
   state2 = true
  }
}
```

If you now want dataflow with this configuration, do: `module Flow = DataFlow::Global<MyConfig>;`. If you want taint flow, do `module Flow = TaintFlow::Global<MyConfig>;`.

To get all the path query machinery, now `import Flow::PathGraph` instead of `import DataFlow::PathGraph` which you used to do.

And in your query, instead of:
```ql
from Configuration config, DataFlow::PathNode source, DataFlow::PathNode sink
where config.hasFlowPath(source, sink)
```
do
```ql
from Flow::PathNode source, Flow::PathNode sink
where Flow::flowPath(source, sink)
```

**Comment by scottconstable** [comment]
Hi @jketema, thank you so much! I was able to get my analysis to work with the legacy taint tracking interface: https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/ir/dataflow/internal/tainttracking1/TaintTrackingImpl.qll/type.TaintTrackingImpl$Configuration.html.

I did have to make one tweak to your suggestion that in "`isAdditionalTaintStep` the two nodes should probably be equal." This was causing several problems for me, most noticeably that the SARIF output was skipping steps in the taint propagation. I fixed it by doing this instead:
``` codeql
override predicate isAdditionalTaintStep(
    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2, DataFlow::FlowState state2
  ) {
    TaintTracking::localTaintStep(node1, node2)
    and
    // characterize the conditions that cause a transition from state1 to state2
  }
```
